Cleanup QueryExecutor and add custom query execution code

jsternberg opened the pull request:
Cleanup QueryExecutor and split statement execution code

The QueryExecutor had a lot of dead code made obsolete by the query
engine refactor that has now been removed. The TSDBStore interface has
also been cleaned up so we can have multiple implementations of this
(such as a local and remote version).

A StatementExecutor interface has been created for adding custom
functionality to the QueryExecutor that may not be available in the open
source version. The QueryExecutor delegate all statement execution to
the StatementExecutor and the QueryExecutor will only keep track of
housekeeping. Implementing additional queries is as simple as wrapping
the cluster.StatementExecutor struct or replacing it with something
completely different.

The PointsWriter in the QueryExecutor has been changed to a simple
interface that implements the one method needed by the query executor.
This is to allow different PointsWriter implementations to be used by
the QueryExecutor. It has also been moved into the StatementExecutor
instead.

The TSDBStore interface has now been modified to contain the code for
creating an IteratorCreator. This is so the underlying TSDBStore can
implement different ways of accessing the underlying shards rather than
always having to access each shard individually (such as batch
requests).

Remove the show servers handling. This isn't a valid command in the open
source version of InfluxDB anymore.

jsternberg commented:
/cc @jwilder @benbjohnson

benbjohnson commented on cluster/query_executor.go:
```diff
@@ -58,6 +60,33 @@ type QueryExecutor struct {
+	// If the plugin executes the statement, it should return true.
+	// Results and errors should be sent over the results channel.
+	ExecuteStatement(stmt influxql.Statement, ctx *ExecutionContext) bool
+}
```
Can this type be named `StatementExecutor` instead? The word "Plugin" doesn't tell much about it and the name fits the function name better.

jsternberg replied:
Yea. I wasn't certain which name to use at the time, but I think `StatementExecutor` is good.

benbjohnson commented on tsdb/store.go:
```diff
@@ -613,6 +614,27 @@ func (s *Store) IteratorCreators() influxql.IteratorCreators {
+	}
+
+	return influxql.IteratorCreators(ics), nil
+}
```
Why is `Store` getting an `IteratorCreator` method?

jsternberg replied:
A couple of reasons. The first is that it meant I could take the `iteratorCreator` method from the `QueryExecutor` and have it be different depending on the underlying store. The `tsdb.Store` is a local store and this will create local iterators from the list of shards. The clustering implementation will do something else in order to ensure we don't make a new connection for each shard.

Then I thought it made sense in this type anyway.

benbjohnson replied:
OK, it seems odd to take a `[]meta.ShardInfo` though if we only need the `ID`. Can you just pass the `[]uint64` shard IDs?

jsternberg replied:
It might be possible, but it would require allocating a new slice and copying the ids into it. I'm not sure it's worth it.

benbjohnson replied:
It's a single allocation in order to make the signature more appropriate. In the grand scheme of query execution I think this is really insignificant.

@jwilder what do you think?

jwilder replied:
Yeah, using `ShardInfo` also creates a dependency on `meta` from `tsdb` package which we've tried really hard to avoid.  `[]uint64` would be more consistent with the other functions in `Store` as well.

jsternberg replied:
I'll change it then.

benbjohnson commented:
This is looking better. Can we axe the `influxql.QueryExecutor` interface and move the `cluster.QueryExecutor` implementation to `influxql`?

jsternberg commented:
It wouldn't be easy for us to do that. The `cluster.QueryExecutor` still uses `cluster.MetaClient` which isn't something that should be in `influxql`. I also like having an interface since I think it helps us with things like testing in the future.

benbjohnson commented:
The only thing we're using the `MetaClient` for is for statement normalization but that should arguably be moved into the `StatementExecutor`.

I generally dislike hiding a single implementation behind an interface. If another package wants to make an interface for `QueryExecutor` for testing then it should live in its own package. The query executor only has one way of executing now that statement execution is delegated.

jsternberg commented:
I've updated the PR to have statement normalization done by the `StatementExecutor` and I've moved things around to merge `QueryManager` into `QueryExecutor` and switch `influxql.QueryExecutor` into a struct instead of an interface.

This should be a lot more clear now that statement execution has been moved out of the `QueryExecutor`.

benbjohnson commented:
:+1:

gunnaraasen commented:
Should the file `cluster/query_executor.go` be renamed to `cluster/statement_executor.go`?

jsternberg commented:
Renamed. It's also failing on CircleCI for some reason but I can't reproduce locally.

jsternberg commented:
I also just renamed the tests to be `QueryExecutor` instead of `QueryManager`. I found a race condition that only happened on go 1.6 that was caught by the tests and fixed it. The CI build should finish correctly now.
